src/auth0-session/session/abstract-session.ts

```diff
@@ -0,0 +1,125 @@
+import createDebug from '../utils/debug';
+import { CookieSerializeOptions } from 'cookie';
+import { Config } from '../config';
+import { Cookies } from '../utils/cookies';
+
+const debug = createDebug('session');
+
+export interface SessionPayload<Session> {
+  header: {
+    /**
+     * Timestamp (in secs) when the session was created.
+     */
+    iat: number;
+    /**
+     * Timestamp (in secs) when the session was last touched.
+     */
+    uat: number;
+    /**
+     * Timestamp (in secs) when the session expires.
+     */
+    exp: number;
+  };
+
+  /**
+   * The session data.
+   */
+  data: Session;
+}
+
+const epoch = (): number => (Date.now() / 1000) | 0; // eslint-disable-line no-bitwise
+export type Header = { iat: number; uat: number; exp: number; [propName: string]: unknown };
+const assert = (bool: boolean, msg: string) => {
+  if (!bool) {
+    throw new Error(msg);
+  }
+};
+
+export abstract class AbstractSession<Req, Res, Session> {
+  constructor(protected config: Config, protected Cookies: new () => Cookies) {}
+
+  abstract getSession(req: Req): Promise<SessionPayload<Session> | undefined | null>;
+
+  abstract setSession(
+    req: Req,
+    res: Res,
+    session: Session,
+    uat: number,
+    iat: number,
+    exp: number,
```

Comment:
If in `getSession()` we're returning a `SessionPayload<Session>`, should `setSession` take a `SessionPayload<Session>` as well?

Reply:
That would makes sense 😄 

But the way the 2 session classes create the header data (one in a JWT header, one in a nested object) - means that it makes more sense for them to create the full payload themselves